### PR TITLE
attach and detach volume by fc protocol

### PR DESCRIPTION
**Pull request by lasshpanzer**
**What this PR does / why we need it**:
* Add Fibre Channel connector support in attach dock, so that any storage driver which supports FC driver could be attached by opensds
* Fix a typo error in Huawei Dorado driver

**Which issue this PR fixes**:

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: contrib/connector/fc/fc.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+// limitations under the License.
+
+package fc
+
+import (
+	"github.com/opensds/opensds/contrib/connector"
+)
+
+var (
+	FC_Driver = "fc"
```
Comment (leonwanghui) [inline contrib/connector/fc/fc.go]: Please change FC_Driver to `FCDriver`

File: contrib/connector/fc/linuxfc.go [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+			hostDevice = fmt.Sprintf("/dev/disk/by-path/pci-%s-fc-%s-lun-%s", pciNum, tgtWWN, processLunId(lun))
+		}
+		hostDevices = append(hostDevices, hostDevice)
+	}
+	return hostDevices
+}
+
+func processLunId(lunId string) string {
+	lunIdInt, _ := strconv.Atoi(lunId)
+	if lunIdInt < 256 {
```
Comment (leonwanghui) [inline contrib/connector/fc/linuxfc.go]: This code style would be better:
```go
	if lunIdInt < 256 {
		return lunId
	}

	return fmt.Sprintf("0x%04x%04x00000000", lunIdInt&0xffff, lunIdInt>>16&0xffff)
}
```

File: contrib/connector/fc/linuxfc.go [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+		return nil, err
+	}
+	var hbasInfos []map[string]string
+	for _, hba := range hbas {
+		wwpn := strings.Replace(hba["prot_name"], "0x", "", -1)
+		wwnn := strings.Replace(hba["node_name"], "0x", "", -1)
+		devicePath := hba["ClassDevicepath"]
+		device := hba["ClassDevice"]
+
+		hbasInfo := map[string]string{"port_name": wwpn, "node_name": wwnn, "host_device": device, "device_path": devicePath}
```
Comment (leonwanghui) [inline contrib/connector/fc/linuxfc.go]: Please update the code style of map struct declaration.

File: contrib/connector/fc/fibreChannel.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+}
+
+func (f *fibreChannel) connectVolume(conn map[string]interface{}) (map[string]string, error) {
+	hbas, err := f.getFChbasInfo()
+	if err != nil {
+		return nil, err
+	}
+	volPaths := f.getVolumePaths(conn, hbas)
+	if len(volPaths) == 0 {
+		errMsg := fmt.Sprintf("No fc devices found.")
```
Comment (xing-yang) [inline contrib/connector/fc/fibreChannel.go]: Please use capital letters for "FC" in the messages.

File: contrib/connector/fc/fibreChannel.go [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+		log.Println(errMsg)
+		return nil, errors.New(errMsg)
+	}
+
+	var tries = 3
+	devicePath, deviceName := f.volPathDiscovery(volPaths, tries, conn["target_wwn"].([]string), hbas)
+	if devicePath != "" && deviceName != "" {
+		log.Printf("Found fibre channel volume name, devicePath is %s, deviceName is %s", devicePath, deviceName)
```
Comment (xing-yang) [inline contrib/connector/fc/fibreChannel.go]: Fibre Channel

File: contrib/connector/fc/fibreChannel.go [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+		return nil, err
+	}
+	volPaths := f.getVolumePaths(conn, hbas)
+	if len(volPaths) == 0 {
+		errMsg := fmt.Sprintf("No fc devices found.")
+		log.Println(errMsg)
+		return nil, errors.New(errMsg)
+	}
+
+	var tries = 3
```
Comment (xing-yang) [inline contrib/connector/fc/fibreChannel.go]: Better to define a constant for "3".

File: contrib/connector/fc/fibreChannel.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+	for i := 0; i < tries; i++ {
+		for _, path := range volPaths {
+			if f.helper.pathExists(path) {
+				deviceName := f.helper.getContentfromSymboliclink(path)
+				return path, deviceName
+			} else {
+				f.helper.rescanHosts(tgtWWN, hbas)
+			}
+		}
+		time.Sleep(2 * time.Second)
```
Comment (xing-yang) [inline contrib/connector/fc/fibreChannel.go]: Better to define a constant for the time.

File: contrib/connector/fc/fibreChannel.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+	if devicePath != "" && deviceName != "" {
+		log.Printf("Found fibre channel volume name, devicePath is %s, deviceName is %s", devicePath, deviceName)
+	}
+
+	deviceWWN, err := f.helper.getSCSIWWN(devicePath)
+	if err != nil {
+		return nil, err
+	}
+
+	return map[string]string{"scsi_wwn": deviceWWN, "path": devicePath}, nil
```
Comment (xing-yang) [inline contrib/connector/fc/fibreChannel.go]: There is no support for multipath?

File: contrib/connector/fc/fibreChannel.go (hunk 6 above)
Comment (lasshpanzer) [reply contrib/connector/fc/fibreChannel.go]: I will change it.

File: contrib/connector/fc/fibreChannel.go (hunk 7 above)
Comment (lasshpanzer) [reply contrib/connector/fc/fibreChannel.go]: In the go grammar rule, this constant can only be used this way. If define a constant for it, there will report a syntax error.

File: contrib/connector/fc/fibreChannel.go (hunk 8 above)
Comment (lasshpanzer) [reply contrib/connector/fc/fibreChannel.go]: I will add multi-path and multi-attach at the same time later. Here i just implemented the basic functions.

**Pull request merged**